arangod/RocksDBEngine/RocksDBVPackIndex.cpp

```diff
@@ -307,143 +487,415 @@ class RocksDBVPackIndexIterator final : public IndexIterator {
-      do {
-        TRI_ASSERT(_index->objectId() ==
-                   RocksDBKey::objectId(_iterator->key()));
+  // look up a value in the in-memory hash cache
+  template<typename F>
+  inline CacheLookupResult lookupInCache(F&& cb, size_t& limit) {
+    TRI_ASSERT(_cache != nullptr);
 
-        --count;
-        ++skipped;
-        if (!advance()) {
-          break;
-        }
+    // key too large to be cached. should almost never happen
+    if (ADB_UNLIKELY(_cacheKeyBuilderSize) >
+        std::numeric_limits<uint32_t>::max()) {
+      return CacheLookupResult::kNotInCache;
+    }
 
-        if (count == 0) {
-          return;
+    size_t const numFields = _index->hasStoredValues() ? 3 : 2;
+    VPackSlice key = _cacheKeyBuilder.slice();
+
+    for (size_t attempts = 0; attempts < 10; ++attempts) {
+      // Try to read from cache
+      auto finding = _cache->find(key.start(),
+                                  static_cast<uint32_t>(_cacheKeyBuilderSize));
+      if (finding.found()) {
+        incrCacheHits();
+        // We got sth. in the cache
+        VPackSlice cachedData(finding.value()->value());
+        TRI_ASSERT(cachedData.isArray());
+        if (cachedData.length() / numFields < limit) {
+          // Directly return it, no need to copy
+          _resultIterator = VPackArrayIterator(cachedData);
+          while (_resultIterator.valid()) {
+            cb();
+            TRI_ASSERT(limit > 0);
+            --limit;
+          }
+          _resultIterator = VPackArrayIterator(VPackArrayIterator::Empty{});
+          return CacheLookupResult::kInCacheAndFullyHandled;
         }
-      } while (true);
-    }
 
-    // validate that Iterator is in a good shape and hasn't failed
-    arangodb::rocksutils::checkIteratorStatus(_iterator.get());
+        // We need to copy the data from the cache, and let the caller
+        // handler the result.
+        _resultBuilder.clear();
+        _resultBuilder.add(cachedData);
+        TRI_ASSERT(_resultBuilder.slice().isArray());
+        _resultIterator = VPackArrayIterator(_resultBuilder.slice());
+        return CacheLookupResult::kInCacheAndPartlyHandled;
+      }  // finding found
+
+      if (finding.result() != TRI_ERROR_LOCK_TIMEOUT) {
+        incrCacheMisses();
+        break;
+      }
+      // enhance our calm...
+      basics::cpu_relax();
+    }  // attempts
+    return CacheLookupResult::kNotInCache;
   }
 
-  /// @brief Reset the cursor
-  void resetImpl() override {
-    TRI_ASSERT(_trx->state()->isRunning());
-    _mustSeek = true;
+  // store a value inside the in-memory hash cache
+  void storeInCache(VPackSlice slice) {
+    TRI_ASSERT(_cache != nullptr);
+
+    // key too large to be cached. should almost never happen
+    if (ADB_UNLIKELY(_cacheKeyBuilderSize) >
+        std::numeric_limits<uint32_t>::max()) {
+      return;
+    }
+
+    size_t attempts = 0;
+    VPackSlice key = _cacheKeyBuilder.slice();
+    cache::Cache::Inserter inserter(
```

Comment:
Essentially the same comment as before: We try 10 times to store the cache value and increase potential contention here 10 times. Is this the right approach?

Reply:
Same here: I reused the approach we currently use for the RocksDBEdgeIndex. But I also think we can remove the `for` loop from the calling code, given that the `insert(...)` call already tries to lock the bucket `triesFast` times.

Reply:
Addressed